python_modules/dagster-graphql/dagster_graphql/schema/auto_materialize_asset_evaluations.py

```diff
@@ -65,23 +70,43 @@ class Meta:
         )
 
 
-def create_graphene_auto_materialize_condition(condition: AutoMaterializeCondition):
-    if isinstance(condition, FreshnessAutoMaterializeCondition):
-        return GrapheneFreshnessAutoMaterializeCondition(decisionType=condition.decision_type)
-    elif isinstance(condition, DownstreamFreshnessAutoMaterializeCondition):
-        return GrapheneDownstreamFreshnessAutoMaterializeCondition(
-            decisionType=condition.decision_type
-        )
-    elif isinstance(condition, ParentMaterializedAutoMaterializeCondition):
-        return GrapheneParentMaterializedAutoMaterializeCondition(
-            decisionType=condition.decision_type
+DAGSTER_NAMEDTUPLE_TO_GRAPHENE_MAP = {
+    FreshnessAutoMaterializeCondition: GrapheneFreshnessAutoMaterializeCondition,
+    DownstreamFreshnessAutoMaterializeCondition: GrapheneDownstreamFreshnessAutoMaterializeCondition,
+    ParentMaterializedAutoMaterializeCondition: GrapheneParentMaterializedAutoMaterializeCondition,
+    MissingAutoMaterializeCondition: GrapheneMissingAutoMaterializeCondition,
+    ParentOutdatedAutoMaterializeCondition: GrapheneParentOutdatedAutoMaterializeCondition,
+}
+
+
+def create_graphene_auto_materialize_condition(
+    condition: Union[AutoMaterializeCondition, Tuple[AutoMaterializeCondition, str]],
+    partitions_def: Optional[PartitionsDefinition],
+):
+    if partitions_def is not None:
+        if not len(condition) == 2 and isinstance(condition[1], str):
+            raise DagsterError(
+                f"Unexpected condition type {type(condition)} for partitions definition"
+                f" {partitions_def}"
+            )
+        auto_materialize_condition, serialized_subset = cast(
+            Tuple[AutoMaterializeCondition, str], condition
         )
-    elif isinstance(condition, MissingAutoMaterializeCondition):
-        return GrapheneMissingAutoMaterializeCondition(decisionType=condition.decision_type)
-    elif isinstance(condition, ParentOutdatedAutoMaterializeCondition):
-        return GrapheneParentOutdatedAutoMaterializeCondition(decisionType=condition.decision_type)
+        subset = partitions_def.deserialize_subset(serialized_subset)
+        partition_keys = subset.get_partition_keys()
     else:
```

Comment:
Imagine the sequence of events:

1. partitioned asset is created
2. auto materializations are kicked off for it, resulting in serialized evaluations
3. that asset's partitions definition is removed

In this codepath, we're using the asset's current partitions definitions to determine what format we're expecting of the serialized data, so we would end up erroring, rather than more gracefully degrading in some way. I leave it to the designers of the world to figure out what exactly we should do there, but ideally in the meantime, users wouldn't see graphql errors in this case. Maybe this function returns None in the case that we detect partitioned data for an asset that does not currently have a partitions definition.

A similar thing should be done for the case that `partitions_def.can_deserialize_subset` is false.

Reply:
With the silly type detection challenges as they are we should just persist an is_partitioned bool on each evaluation

And yeah it should handle cases where it can't deserialize

Reply:
I think we ran into a similar issue with deserializing asset backfills in the past.

The solution we landed on was to serialize (a) the partitions def serializable unique ID and (b) class name for the partitions subset for each partitions subset. This enables:
- detecting when a partitions def has been removed.
- for time window partitions subsets, checking the serializable unique ID to see if the partitions def is still the same. If it is not, we can no longer deserialize.
- for other partitions subsets (static/multipartitioned), if the class name is equivalent, we can deserialize since it just stores a set of partition keys.

This is all logic that already exists in `partitions_def.can_deserialize_subset`. I suspect we'll likely want to implement something similar for these cases, though it's up to you on if you want to implement this now or in the future.

Reply:
Added this via the SerializedPartitionsSubset